lib/taskana-core/src/main/java/pro/taskana/jobs/TaskCleanupJob.java

```diff
@@ -61,10 +65,44 @@ public void run() throws TaskanaException {
     }
 
     private List<TaskSummary> getTasksCompletedBefore(Instant untilDate) {
-        return taskanaEngineImpl.getTaskService()
-            .createTaskQuery()
-            .completedWithin(new TimeInterval(null, untilDate))
-            .list();
+        List<TaskSummary> taskList = taskanaEngineImpl.getTaskService()
+                .createTaskQuery()
+                .completedWithin(new TimeInterval(null, untilDate))
+                .orderByBusinessProcessId(asc)
+                .list();
+
+        if (allCompletedSameParentBusiness) {
+            Map<String, Long> numberParentTasksShouldHave = new HashMap<>();
+            Map<String, Long> countParentTask = new HashMap<>();
+            for (TaskSummary task : taskList) {
+                numberParentTasksShouldHave.put(task.getParentBusinessProcessId(), taskanaEngineImpl.getTaskService().createTaskQuery().parentBusinessProcessIdIn(task.getParentBusinessProcessId()).count());
+                if (countParentTask.get(task.getParentBusinessProcessId()) != null) {
```

Comment:
this is way more readable: `countParentTask.merge(task.getParentBusinessProcessId(), 1L, Long::sum);`